pkg/catalogv2/oci/oci.go

```diff
@@ -0,0 +1,234 @@
+					ociClient.repository = repository
+					ociURL := fmt.Sprintf("%s/%s", ociClient.registry, ociClient.repository)
+
+					orasRepository, err = ociClient.GetOrasRepository()
+					if err != nil {
+						return fmt.Errorf("failed to create an oras repository for url %s: %w", ociURL, err)
+					}
+
+					err = orasRepository.Tags(context.Background(), "", tagsFunc)
+					if err != nil {
+						return fmt.Errorf("failed to fetch tags for repository %s: %w", ociURL, err)
+					}
+				}
+			}
+			ociClient.repository = userProvidedRepository
+		}
+		return nil
+	}
+
+	IsOrasRepository, err := ociClient.IsOrasRepository()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the user has provided the tag, we simply generate the index with that single oci artifact.
+	if ociClient.tag != "" {
+		orasRepository, err = ociClient.GetOrasRepository()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create an oras repository for url %s: %w", URL, err)
+		}
+
+		err = addToHelmRepoIndex(*ociClient, chartName, indexFile, orasRepository)
+		if err != nil {
+			return nil, fmt.Errorf("failed to add chartName %s in oci URL %s to helm repo index: %w", chartName, URL, err)
+		}
+
+		// If the repository is provided with no tag, then we fetch all tags.
+	} else if IsOrasRepository {
+		orasRepository, err = ociClient.GetOrasRepository()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create an oras repository for url %s: %w", URL, err)
+		}
+
+		err = orasRepository.Tags(context.Background(), "", tagsFunc)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch tags for repository %s: %w", URL, err)
+		}
+
+		// If no repository and tag is provided, we fetch
+		// all repositories and then tags associated.
+	} else {
+		orasRegistry, err := ociClient.GetOrasRegistry()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create an oras registry for %s: %w", ociClient.registry, err)
+		}
+
+		// Fetch all repositories
+		err = orasRegistry.Repositories(context.Background(), "", repositoriesFunc)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch repositories for %s: %w", URL, err)
+		}
+	}
+
+	return indexFile, nil
+}
+
+// addToHelmRepoIndex adds the helmchart aka oras repository to the helm repo index
+func addToHelmRepoIndex(ociClient Client, chartName string, indexFile *repo.IndexFile, orasRepository *remote.Repository) error {
+	ociURL := fmt.Sprintf("%s/%s:%s", ociClient.registry, ociClient.repository, ociClient.tag)
+
+	// Fetch the helm chart
+	filePath, isHelmChart, err := ociClient.fetchChart(orasRepository)
+	if err != nil {
+		return fmt.Errorf("failed to fetch an helm chart %s: %w", ociURL, err)
+	}
+
+	// Add to the index if it is a helm chart
+	if isHelmChart {
+		err = ociClient.addToIndex(indexFile, chartName, filePath)
+		if err != nil {
+			return fmt.Errorf("unable to add helm chart %s to index: %w", ociURL, err)
+		}
+	}
+
+	// Delete the file
+	err = os.Remove(filePath)
```

Comment:
If `isHelmChart` is `false`, this files would accumulate in disk without being removed - potentially causing disk pressure.
Please move this to line `218` with a `defer`.

Reply:
Done.